mm/pmm.c

```diff
@@ -41,11 +41,28 @@ static frames_array_t early_frames;
 static list_head_t free_frames[MAX_PAGE_ORDER + 1];
 static list_head_t busy_frames[MAX_PAGE_ORDER + 1];
 
-#define MIN_NUM_4K_FRAMES 2
+/* 3 for paging and 4 for refilling (1 array frame, 3 for mapping it) */
+#define MIN_NUM_4K_FRAMES 3 + (1 + 3)
+/* enough frames to refill the 4K frames in worst case without running our of space */
+#define MIN_NOREFILL_FREE_FRAMES_THRESHOLD                                               \
+    (MIN_FREE_FRAMES_THRESHOLD + MIN_NUM_4K_FRAMES + (MAX_PAGE_ORDER - PAGE_ORDER_4K))
 static size_t frames_count[MAX_PAGE_ORDER + 1];
 
+static bool refilling;
 static spinlock_t lock = SPINLOCK_INIT;
 
+static void try_create_4k_frames(void);
+
+static void spin_lock_norefill() {
+    spin_lock(&lock);
+
+    while (refilling) {
+        spin_unlock(&lock);
+        /* TODO: some backoff would be nice here... */
+        spin_lock(&lock);
+    }
```

Comment:
This code is unsettling. Feels like an antipattern to me. 

You correctly identified the problem being with roundtrips between VMM and PMM due to allocation of new array frames for PMM. Best would be to get rid of it entirely, but we want to preserve the dynamic nature of PMM frames allocation. Therefore, owner of the locks (the initiator of the operation) needs to be able to finish with all the roundtrips, while new operations (by other CPUs) wait on the locks. Indeed, looks like a job for a mutex of some sort.

But maybe (I have not had time to look closely into this) it is possible to add non-locking internal interfaces to both VMM and PMM and use them for the problematic roundtrips. Assumption to check would be if the locks already held at the start of the operation are sufficient protection against other CPUs.

Reply:
Yeah this one is horrible :sweat_smile: . I now try to solve it using two-stage locking instead.